File: bwapi-bridge/etc/bwsal-src/cpp/BuildManager.cpp
#include <BuildManager.h>
#include <BuildingPlacer.h>
#include <ConstructionManager.h>
#include <ProductionManager.h>
#include <MorphManager.h>
BuildManager::BuildManager(Arbitrator::Arbitrator<BWAPI::Unit*,double>* arbitrator)
{
  this->arbitrator=arbitrator;
  this->buildingPlacer=new BuildingPlacer();
  this->constructionManager=new ConstructionManager(this->arbitrator,this->buildingPlacer);
  this->productionManager=new ProductionManager(this->arbitrator,this->buildingPlacer);
  this->morphManager=new MorphManager(this->arbitrator);
}

BuildManager::~BuildManager()
{
  delete this->buildingPlacer;
  delete this->constructionManager;
  delete this->productionManager;
  delete this->morphManager;
}

void BuildManager::update()
{
  this->constructionManager->update();
  this->productionManager->update();
  this->morphManager->update();
}

std::string BuildManager::getName() const
{
  return "Build Manager";
}

BuildingPlacer* BuildManager::getBuildingPlacer() const
{
  return this->buildingPlacer;
}
void BuildManager::onRemoveUnit(BWAPI::Unit* unit)
{
  this->constructionManager->onRemoveUnit(unit);
  this->productionManager->onRemoveUnit(unit);
  this->morphManager->onRemoveUnit(unit);
}

bool BuildManager::build(BWAPI::UnitType type)
{
  return build(type, BWAPI::Broodwar->self()->getStartLocation());
}

bool BuildManager::build(BWAPI::UnitType type, BWAPI::TilePosition goalPosition)
{
  if (type==BWAPI::UnitTypes::None || type==BWAPI::UnitTypes::Unknown) return false;
  if (type.getRace()==BWAPI::Races::Zerg && type.isBuilding()==type.whatBuilds().first->isBuilding())
    return this->morphManager->morph(type);
  else
  {
    if (type.isBuilding())
      return this->constructionManager->build(type, goalPosition);
    else
      return this->productionManager->train(type);
  }
  return false;
}

int BuildManager::getPlannedCount(BWAPI::UnitType type) const
{
  if (type.getRace()==BWAPI::Races::Zerg && type.isBuilding()==type.whatBuilds().first->isBuilding())
    return BWAPI::Broodwar->self()->completedUnitCount(type)+this->morphManager->getPlannedCount(type);
  else
  {
    if (type.isBuilding())
      return BWAPI::Broodwar->self()->completedUnitCount(type)+this->constructionManager->getPlannedCount(type);
    else
      return BWAPI::Broodwar->self()->completedUnitCount(type)+this->productionManager->getPlannedCount(type);
  }
}

int BuildManager::getStartedCount(BWAPI::UnitType type) const
{
  if (type.getRace()==BWAPI::Races::Zerg && type.isBuilding()==type.whatBuilds().first->isBuilding())
    return BWAPI::Broodwar->self()->completedUnitCount(type)+this->morphManager->getStartedCount(type);
  else
  {
    if (type.isBuilding())
      return BWAPI::Broodwar->self()->completedUnitCount(type)+this->constructionManager->getStartedCount(type);
    else
      return BWAPI::Broodwar->self()->completedUnitCount(type)+this->productionManager->getStartedCount(type);
  }
}

int BuildManager::getCompletedCount(BWAPI::UnitType type) const
{
  return BWAPI::Broodwar->self()->completedUnitCount(type);
}

void BuildManager::setBuildDistance(int distance)
{
  this->buildingPlacer->setBuildDistance(distance);
}

File: bwapi-bridge/src/org/bwapi/bridge/sal/BuildManager.java
package org.bwapi.bridge.sal;

import org.bwapi.bridge.model.Game;
import org.bwapi.bridge.model.Race;
import org.bwapi.bridge.model.TilePosition;
import org.bwapi.bridge.model.Unit;
import org.bwapi.bridge.model.UnitType;


/**
 * Build manager
 * 
 * @see <a href="http://code.google.com/p/bwsal/wiki/BuildManager">BuildManager</a>
 * 
 * @author Chad Retz
 *
 * @since 0.3
 */
public class BuildManager {
    
    //TODO: why is this here?
    protected final Arbitrator<Unit, Double> arbitrator;
    protected final BuildingPlacer buildingPlacer;
    protected final ConstructionManager constructionManager;
    protected final ProductionManager productionManager;
    protected final MorphManager morphManager;
    
    public BuildManager(Arbitrator<Unit, Double> arbitrator) {
        this.arbitrator = arbitrator;
        buildingPlacer = new BuildingPlacer();
        constructionManager = new ConstructionManager(arbitrator, buildingPlacer);
        productionManager = new ProductionManager(arbitrator, buildingPlacer);
        morphManager = new MorphManager(arbitrator);
    }
    
    public void update() {
        constructionManager.update();
        productionManager.update();
        morphManager.update();
    }
    
    public String getName() {
        return "Build Manager";
    }
    
    public BuildingPlacer getBuildingPlacer() {
        return buildingPlacer;
    }
    
    public void onRemoveUnit(Unit unit) {
        constructionManager.onRemoveUnit(unit);
        productionManager.onRemoveUnit(unit);
        morphManager.onRemoveUnit(unit);
    }

    public boolean build(UnitType type) {
        return build(type, Game.getInstance().self().getStartLocation());
    }
    
    public boolean build(UnitType type, TilePosition goalPosition) {
        if (type.equals(UnitType.NONE) || type.equals(UnitType.UNKNOWN)) {
            return false;
        }
        if (type.getRace().equals(Race.ZERG) && type.isBuilding() == type.
                whatBuilds().getKey().isBuilding()) {
            return morphManager.morph(type);
        } else if (type.isBuilding()) {
            return constructionManager.build(type, goalPosition);
        } else {
            return productionManager.train(type);
        }
    }
    
    public int getPlannedCount(UnitType type) {
        if (type.getRace().equals(Race.ZERG) && type.isBuilding() == 
                type.whatBuilds().getKey().isBuilding()) {
            return Game.getInstance().self().completedUnitCount(type) +
                morphManager.getPlannedCount(type);
        } else if (type.isBuilding()) {
            return Game.getInstance().self().completedUnitCount(type) +
                constructionManager.getPlannedCount(type);
        } else {
            return Game.getInstance().self().completedUnitCount(type) +
                productionManager.getPlannedCount(type);
        }
    }
    
    public int getStartedCount(UnitType type) {
        if (type.getRace().equals(Race.ZERG) && type.isBuilding() == 
                type.whatBuilds().getKey().isBuilding()) {
            return Game.getInstance().self().completedUnitCount(type) +
                morphManager.getStartedCount(type);
        } else if (type.isBuilding()) {
            return Game.getInstance().self().completedUnitCount(type) +
                constructionManager.getStartedCount(type);
        } else {
            return Game.getInstance().self().completedUnitCount(type) +
                productionManager.getStartedCount(type);
        }
    }

    public int getCompletedCount(UnitType type) {
        return Game.getInstance().self().completedUnitCount(type);
    }
    
    public void setBuildDistance(int distance) {
        buildingPlacer.setBuildDistance(distance);
    }
}
